go.mod

```diff
@@ -1,43 +1,41 @@
 module github.com/SAP/stewardci-core
 
-go 1.12
+go 1.14
 
 require (
-	cloud.google.com/go v0.48.0 // indirect
-	contrib.go.opencensus.io/exporter/prometheus v0.1.0 // indirect
-	contrib.go.opencensus.io/exporter/stackdriver v0.12.8 // indirect
+	github.com/Azure/go-autorest/autorest v0.10.2 // indirect
+	github.com/Azure/go-autorest/autorest/adal v0.8.3 // indirect
+	github.com/aws/aws-sdk-go v1.31.12 // indirect
+	github.com/containerd/containerd v1.3.4 // indirect
 	github.com/davecgh/go-spew v1.1.1
-	github.com/evanphx/json-patch v4.5.0+incompatible // indirect
 	github.com/ghodss/yaml v1.0.0
-	github.com/gogo/protobuf v1.3.1 // indirect
-	github.com/golang/groupcache v0.0.0-20191027212112-611e8accdfc9 // indirect
-	github.com/golang/mock v1.3.1
-	github.com/google/go-containerregistry v0.0.0-20191115225042-f8574ec722f4 // indirect
+	github.com/golang/mock v1.4.3
+	github.com/golang/protobuf v1.4.2 // indirect
+	github.com/google/go-cmp v0.4.1 // indirect
 	github.com/google/uuid v1.1.1
-	github.com/googleapis/gnostic v0.3.1 // indirect
-	github.com/hashicorp/golang-lru v0.5.3 // indirect
-	github.com/imdario/mergo v0.3.8 // indirect
+	github.com/imdario/mergo v0.3.9 // indirect
+	github.com/konsorten/go-windows-terminal-sequences v1.0.3 // indirect
 	github.com/lithammer/dedent v1.1.0
-	github.com/markbates/inflect v1.0.4 // indirect
-	github.com/mattbaird/jsonpatch v0.0.0-20171005235357-81af80346b1a // indirect
-	github.com/openzipkin/zipkin-go v0.2.2 // indirect
-	github.com/pkg/errors v0.8.1
-	github.com/prometheus/client_golang v1.2.1
-	github.com/tektoncd/pipeline v0.8.0
-	go.uber.org/zap v1.13.0 // indirect
+	github.com/opencontainers/go-digest v1.0.0 // indirect
+	github.com/pkg/errors v0.9.1
+	github.com/prometheus/client_golang v1.6.0
+	github.com/tektoncd/pipeline v0.10.2
+	google.golang.org/appengine v1.6.6 // indirect
+	google.golang.org/grpc v1.29.1 // indirect
 	gotest.tools v2.2.0+incompatible
-	k8s.io/api v0.0.0-20180904230853-4e7be11eab3f
-	k8s.io/apimachinery v0.0.0-20180904193909-def12e63c512
-	k8s.io/client-go v0.0.0-20180910083459-2cefa64ff137
-	k8s.io/utils v0.0.0-20191114200735-6ca3b61696b6 // indirect
-	knative.dev/pkg v0.0.0-20191107185656-884d50f09454
+	k8s.io/api v0.17.6
+	k8s.io/apimachinery v0.17.6
+	k8s.io/client-go v11.0.1-0.20190805182717-6502b5e7b1b5+incompatible
```

Comment:
Can't/shouldn't we update to `v0.16.10` (`kubernetes-1.16.10`) since K8S_VERSION is also updated to 1.16.10?
See also [compatibility-matrix](https://github.com/kubernetes/client-go#compatibility-matrix)

Reply:
I want `go.mod` to be the result of `go mod tidy`, so that by running `go mod tidy` again no changes are introduced.

`go mod tidy` always enters the result of [minimal version selection](https://github.com/golang/go/wiki/Modules#version-selection). In our case the minimum required versions are higher than the version we want to use in our build. This is why we have the respective `replace` statements.

`go mod list -m` shows this:
```
...
k8s.io/api v0.17.6 => k8s.io/api v0.16.10
k8s.io/apiextensions-apiserver v0.17.2
k8s.io/apimachinery v0.17.6 => k8s.io/apimachinery v0.16.10
k8s.io/apiserver v0.17.6
k8s.io/cli-runtime v0.17.3
k8s.io/client-go v11.0.1-0.20190805182717-6502b5e7b1b5+incompatible => k8s.io/client-go v0.16.10
k8s.io/cloud-provider v0.17.6
k8s.io/code-generator v0.18.0
k8s.io/component-base v0.17.6
k8s.io/csi-translation-lib v0.17.6
k8s.io/gengo v0.0.0-20200205140755-e0e292d8aa12
k8s.io/klog v1.0.0
k8s.io/kube-openapi v0.0.0-20200410163147-594e756bea31
k8s.io/kubectl v0.17.2
k8s.io/kubernetes v1.13.0
k8s.io/legacy-cloud-providers v0.17.6
k8s.io/metrics v0.17.2
k8s.io/test-infra v0.0.0-20200407001919-bc7f71ef65b8
k8s.io/utils v0.0.0-20191114184206-e782cd3c129f
knative.dev/pkg v0.0.0-20200528142800-1c6815d7e4c9 => knative.dev/pkg v0.0.0-20200528142800-1c6815d7e4c9
...
```

`k8s.io/client-go` is a special thing. It has been using v2+ Git tags long before opting in to modules (version of client-go). For modules they introduced new Git tags v0.<k8s_minor>.<k8s_patch> (Kubernetes version). Now the repo contains two separate versioning tag sets, both adhering to semver. The Go dependency resolution cannot distinguish them. A v11 pseudo version (client-go version) is semantically higher that a v0 version (K8s version), so it gets chosen by the minimal version selection algorithm. But actually the comparison is nonsense.

Reply:
Do we have to use `go mod tidy`? Or can't (shouldn't) we use the explicitely specified versions in the central build?

Reply:
> Do we have to use `go mod tidy`?

`go mod tidy` is _the_ tool to create the canonical form of `go.mod`, including the removal of unused require statements. And in contrast to other go subcommands it considers _any_ build configuration, not just one like `go list` or `go build`.

> Or can't (shouldn't) we use the explicitely specified versions in the central build?

We use `replace` statements to define _certain_ module versions explicitly. All others are the result of minimal version selection among all dependencies. We certainly do _not_ want to set _all_ versions explicitly, as this can easily create non-working version combinations. By the way, with go modules the dependency resolution is 100% reproducible (same result for same `go.mod`).